### ISSUE TITLE
Image Rendering

### ISSUE THREAD

**Issue by christianparpart** [christianparpart/contour#100]
The idea is, to not mess around with the main/alt textual buffer for displaying graphics, but also support switching to a graphics buffer that purely supports rendering graphics (such as images).

The app could use the already existing `CSI 14 t` to retrieve pixel dimensions, programmatically feature-test if graphics backbuffer is supported, then switch to it, and then upload the to-be-displayed image in at least the following way(s).

# Checklist

* [ ] `CSI 14 t` must be supported (get screen dimension in pixels)
* [ ] terminal emulator identification via VT sequence? (so apps can test that it is supported).
* [ ] inline image rendering in main/alt buffer (should still be supported)
* [ ] graphics backbuffer enabled image rendering.
* [ ] inline image rendering via sixels? (this is the old xterm way?)

# References
* [Neofetch](https://github.com/dylanaraps/neofetch) is a nice tool to test inline images
* https://github.com/dylanaraps/neofetch/wiki/Images-in-the-terminal

**Comment by data-man** [comment]
@christianparpart 
Some terminal image viewers supports iTerm2 output.
If you don't mind I want to add this feature to Contour. :)

**Comment by christianparpart** [comment]
> @christianparpart Some terminal image viewers supports iTerm2 output. If you don't mind I want to add this feature to Contour. :)

While I'd really love to have progress on image rendering in especially Contour, I do not think that adding iTerm2's image protocol is the right move here - just because some apps have support for it.

If I wouldn't be overcharged right now, I'd continue working on https://github.com/contour-terminal/terminal-good-image-protocol and push for it (will do that as soon as possible). But I am actually really happy to see that you have interest in coding parts of Contour, too. Maybe we can also pick another topic that would not bore you?

**Comment by data-man** [comment]
> I do not think that adding iTerm2's image protocol is the right move here

That's a pity.

> Maybe we can also pick another topic that would not bore you?

I'd be interested in working on libunicode. :)

**Comment by timsofteng** [comment]
Hello. Any updates about it?

**Comment by christianparpart** [comment]
Hey. I keep you posted. We currently focus on moving/upgrading the frontend from Qt5 to Qt6/QML. I certainly want to have the GIP PR merged this year, too.

Sorry if it takes all a little long, but we are limited on resources. If you see fit where you can help us (anywhere anyhow), if you want, feel free to join us on Discord. 

Take care,
Christian.

**Comment by christianparpart** [comment]
I agree. I personally have no issues with Sixel myself, well, until last week, where I wanted to have an image *behind* the text.
I personally think however, that KGP is way to complex, and thus, other TEs just keep implementing subsets of it. so yeah. I'm not pushing you in either direction. Thanks for answering though. :)

**Comment by christianparpart** [comment]
> Extension 11 in the DA1 report is already assigned to the status display extension. Again you need to pick a number that isn't already used. Avoid anything below 50.

@j4james I was searching on the net about this. Quite funnily, we do support Status Display, so it wouldn't be wrong all along. I've changed that to `60` for this now though. However, for Status Display being advertised as `DA1`'s `11` I cannot find anything on the net telling me that this is the actual number. Where did you get this information from? I'd love to add DA1 11 though, since we do have status / host writable display.

**Comment by j4james** [comment]
You can find a list of most extension numbers in the [Terminal Management chapter of the DEC VSRM](https://j4james.github.io/vtdocs/EL-00070-04.html#4-19). There are a few more found in the programmer references for later DEC devices like the [VT510](https://vt100.net/docs/vt510-rm/DA1.html), and some other supplementary docs.

But note that you don't technically need to report the status extension if you're claiming level 5 conformance (that's what the `65` indicates at the start of your `DA1` report), because it's a _required_ extension at level 3 and above - any terminal that report level 5 conformance is _expected_ to support the status extension.

The real issue is that Contour shouldn't be claiming level 5 conformance in the first place, because it doesn't meet all of the conformance requirements. It's impossible for applications to tell what functionality you _do_ actually support if your `DA1` report indicates an incorrect conformance level.

I know the documentation on this subject is difficult to figure out, because it's spread all over the place. If in doubt, though, it's best to report a low conformance level, and then include explicit extension numbers for the extensions that Contour _does_ actually support. For the most part the higher conformance levels are just sets of required extensions in order to keep the reported extension list shorter.

**Comment by christianparpart** [comment]
Nice, many thanks, the link to your page was actually searching for. Thanks for that.

Regarding level 5 conformance. I'd rather push for implementing what is currently the gap rather than calling defeat (assuming it won't take a year). However, I think not everything makes sense in todays world either. 🤔

**Comment by j4james** [comment]
Obviously it would be great if you could eventually support everything, but right now Contour will fail to work with some applications entirely because of its incorrect `DA1` report. As one example, an application may choose to use macros for optimisation, but fallback to a slower output format on terminals that don't support that extension. However, if Contour is incorrectly claiming macro support, it simply won't work - the application won't know that it should be using the fallback because it doesn't know that is needed on Contour.
